src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,181 @@
+            // side will return immediately, without actually notifying the project system.  That
+            // means that we could persist the data to local storage that isn't in sync with what
+            // the project system knows about.  i.e. if VS is closed or crashes before that
+            // information is persisted, then these two systems will be in disagreement.  this is
+            // believed to not be a big issue given how small a time window this would be and how
+            // easy it would be to get out of that state (just edit the file).
+
+            await PersistLatestInfosAsync(projectVersion, latestInfos, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task PersistLatestInfosAsync(VersionStamp projectVersion, (Document, DesignerInfo? info, bool changed)[] latestInfos, CancellationToken cancellationToken)
+        {
+            foreach (var (doc, info, _) in latestInfos)
+            {
+                // Skip documents that didn't change contents/version at all.  No point in writing
+                // back out the exact same data as before.
+                if (info == null)
+                    continue;
+
+                using var memoryStream = new MemoryStream();
+                using var writer = new ObjectWriter(memoryStream);
+
+                info.Value.WriteTo(writer, projectVersion);
+
+                memoryStream.Position = 0;
+                await _storage.WriteStreamAsync(
+                    doc, DataKey, memoryStream, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<(Document, DesignerInfo? info, bool changed)[]> ComputeLatestInfosAsync(
+            Project project, VersionStamp projectVersion,
+            Document? specificDoc, CancellationToken cancellationToken)
+        {
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var designerCategoryType = compilation.DesignerCategoryAttributeType();
+
+            using var _1 = ArrayBuilder<Task<(Document, DesignerInfo?, bool changed)>>.GetInstance(out var tasks);
+            foreach (var document in project.Documents)
+            {
+                // If we're only analyzing a specific document, then skip the rest.
+                if (specificDoc != null && document != specificDoc)
+                    continue;
+
+                tasks.Add(ComputeDesignerAttributeInfoAsync(
+                    projectVersion, designerCategoryType, document, cancellationToken));
+            }
+
+            var latestInfos = await Task.WhenAll(tasks).ConfigureAwait(false);
+            return latestInfos;
+        }
+
+        private async Task<(Document, DesignerInfo?, bool changed)> ComputeDesignerAttributeInfoAsync(
+            VersionStamp projectVersion, INamedTypeSymbol? designerCategoryType,
+            Document document, CancellationToken cancellationToken)
+        {
```

Comment:
>{ [](start = 8, length = 1)

Wrap this into `try { } catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCancelled(e)) { return default; }`. 
Since these tasks are executed by `Task.WhenAll` the stack frames won't be preserved in crash dump if this code throws.

Reply:
sure

Reply:
done.

Reply:
done.